pkg/controllers/infracluster/infracluster_controller.go

```diff
@@ -0,0 +1,288 @@
+	case configv1.PowerVSPlatformType:
+		log.Info(fmt.Sprintf("detected platform %q is supported, but not implemented yet", r.Platform))
+	case configv1.VSpherePlatformType:
+		log.Info(fmt.Sprintf("detected platform %q is supported, but not implemented yet", r.Platform))
+	case configv1.OpenStackPlatformType:
+		log.Info(fmt.Sprintf("detected platform %q is supported, but not implemented yet", r.Platform))
+	default:
+		log.Info(fmt.Sprintf("detected platform %q is not supported, skipping capi controllers setup", r.Platform))
+	}
+
+	// At this point, the InfraCluster exists.
+	managedByAnnotationVal, found := r.InfraCluster.GetAnnotations()[clusterv1.ManagedByAnnotation]
+	if !found {
+		// Could not find the managedByAnnotation on the InfraCluster object.
+		// This means, by definition, that the object is directly managed by CAPI.
+		// No action should be taken by this controller.
+		return ctrl.Result{}, nil
+	}
+
+	switch managedByAnnotationVal {
+	case managedByAnnotationValueClusterCAPIOperator:
+		// At this point it is this controller's responsibility to manage this InfraCluster object.
+		isReady, err := getReadiness(r.InfraCluster)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("unable to get readiness for InfraCluster: %w", err)
+		}
+		if isReady {
+			// The Infrastructure for this CAPI Cluster is already ready - nothing to do.
+			return ctrl.Result{}, nil
+		}
+
+		infraClusterCopy, ok := r.InfraCluster.DeepCopyObject().(client.Object)
+		if !ok {
+			return ctrl.Result{}, fmt.Errorf("unable to convert InfraCluster runtime.Object to client.Object")
+		}
+
+		infraClusterPatchCopy := r.InfraCluster.DeepCopyObject().(client.Object)
+		if !ok {
+			return ctrl.Result{}, fmt.Errorf("unable to convert InfraCluster runtime.Object to client.Object")
+		}
+
+		// Set Status.Ready=true to indicate that cluster's infrastructure ready.
+		if err := setReadiness(infraClusterCopy, true); err != nil {
+			return ctrl.Result{}, fmt.Errorf("unable to set readiness for InfraCluster: %v", err)
+		}
+
+		if err := r.Client.Status().Patch(ctx, infraClusterCopy, client.MergeFrom(infraClusterPatchCopy)); err != nil {
+			return ctrl.Result{}, fmt.Errorf("unable to patch InfraCluster: %v", err)
+		}
+	default:
+		// At this point it is not this controller's responsibility to manage this InfraCluster object, nor is it
+		// the CAPI infra providers responsbility to do so. This means this object was created outside of these two entities - thus
+		// the creating entity must manage its readiness.
+		log.Info(fmt.Sprintf("InfraCluster is annotated with an unrecognized externally managed annotation value %q"+
+			"- skipping as not managed by this controller.", managedByAnnotationVal))
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// ensureAWSCluster ensures the AWSCluster cluster object exists.
+func (r *InfraClusterController) ensureAWSCluster(ctx context.Context) error {
+	target := awsv1.AWSCluster{ObjectMeta: metav1.ObjectMeta{
```

Comment:
Typically you declare as a pointer here rather than taking the address several times through this function